### PR TITLE
Lower default relay fees

### PR DESCRIPTION
**Pull request by ajtowns**
In the meeting of [2018-07-05](http://www.erisian.com.au/meetbot/bitcoin-core-dev/2018/bitcoin-core-dev.2018-07-05-19.22.log.html#l-24), we discussed dropping the minimum fee rate below 1000 satoshi/kB. This patch set does only that, leaving related features for other PRs.

 * it changes the min tx fee to 200 s/kvB, and the incremental relay fee to 100 s/B
 * it leaves wallet min fee and incremental fee at 1000 s/kvB and 5000 s/kvB for compatibility while the network upgrades
 * it changes the functional tests that are written assuming 1000s/kvB as the min fee levels to explicitly set that via command line parameter

There's a bit more of an explanation for some of the choices in the individual commit messages.

Related PRs:

 * #13987 -- report minfee in getpeerinfo. useful to observe if your peers have lowered their min relay fees
 * #13988 -- settxfee checks. this is useful since currently settxfee to the minrelaytxfee works fine, but with a lower minrelaytxfee, settxfee will be ignored for the range of values between minrelaytxfee and mintxfee which can be confusing
 * #13990 -- fix fee estimation to work with lower fees; currently fee estimation won't give estimates lower than 1000 satoshi per kilobyte.

### CODE REVIEW

**Comment by ken2812221** [comment]
Concept ACK

File: src/policy/fees.cpp [hunk 1]
```diff
@@ -195,6 +202,45 @@ TxConfirmStats::TxConfirmStats(const std::vector<double>& defaultBuckets,
...
+    decay = txstatsSource.decay;
+    scale = txstatsSource.scale;
+
+    const unsigned int maxPeriods = txstatsSource.confAvg.size();
+    assert(txstatsSource.failAvg.size() == maxPeriods);
+
+    confAvg.resize(maxPeriods);
+    for (unsigned int i = 0; i < maxPeriods; i++) {
+        confAvg[i].resize(buckets.size());
+    }
```
Comment (domob1812) [inline src/policy/fees.cpp]: Nit (here and below):  You could use a C++11 for-each loop.

File: src/test/mempool_tests.cpp [hunk 2]
```diff
@@ -532,43 +532,43 @@ BOOST_AUTO_TEST_CASE(MempoolSizeLimitTest)
...
-    // ... but feerate should never drop below 1000
-
     SetMockTime(42 + 8*CTxMemPool::ROLLING_FEE_HALFLIFE + CTxMemPool::ROLLING_FEE_HALFLIFE/2 + CTxMemPool::ROLLING_FEE_HALFLIFE/4);
+    BOOST_CHECK_EQUAL(pool.GetMinFee(1).GetFeePerK(), 100);
+    // ... but feerate should never drop below the min incremental relay fee
+
+    SetMockTime(42 + 9*CTxMemPool::ROLLING_FEE_HALFLIFE + CTxMemPool::ROLLING_FEE_HALFLIFE/2 + CTxMemPool::ROLLING_FEE_HALFLIFE/4);
     BOOST_CHECK_EQUAL(pool.GetMinFee(1).GetFeePerK(), 0);
-    // ... unless it has gone all the way to 0 (after getting past 1000/2)
+    // ... unless it has gone all the way to 0 (after getting past minrelayfee/2)
```
Comment (MarcoFalke) [inline src/test/mempool_tests.cpp]: Instead of changing all the constants in this test, couldn't you set the global min relay fee at the start of the test to 1000 and then back to what it was initially at the end of the test?

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -2982,6 +2982,14 @@ static UniValue settxfee(const JSONRPCRequest& request)
     LOCK2(cs_main, pwallet->cs_wallet);
 
     CAmount nAmount = AmountFromValue(request.params[0]);
+    CFeeRate nFeeRate(nAmount, 1000);
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp]: nit: Should be snake_case: `tx_fee_rate`.

Also, could submit this feature independent of this pull request, i.e. create a separate pull request.

File: src/rpc/net.cpp [hunk 4]
```diff
@@ -169,6 +170,7 @@ static UniValue getpeerinfo(const JSONRPCRequest& request)
             obj.pushKV("inflight", heights);
         }
         obj.pushKV("whitelisted", stats.fWhitelisted);
+        obj.pushKV("minfee", ValueFromAmount(stats.minFeeFilter));
```
Comment (MarcoFalke) [inline src/rpc/net.cpp]: Missing test. Also, this feature should probably be submitted in separate pull request.

File: src/test/mempool_tests.cpp (hunk 2 above)
Comment (ajtowns) [reply src/test/mempool_tests.cpp]: I thought it made more sense to check the default behaviour works right, though that would be plausible too

File: src/policy/fees.cpp [hunk 5]
```diff
@@ -195,6 +202,45 @@ TxConfirmStats::TxConfirmStats(const std::vector<double>& defaultBuckets,
     resizeInMemoryCounters(buckets.size());
 }
 
+TxConfirmStats::TxConfirmStats(const std::vector<double>& defaultBuckets, const std::map<double, unsigned int>& defaultBucketMap, const TxConfirmStats& txstatsSource)
+    : buckets(defaultBuckets), bucketMap(defaultBucketMap)
+{
+    decay = txstatsSource.decay;
+    scale = txstatsSource.scale;
```
Comment (Empact) [inline src/policy/fees.cpp]: nit: these can go in the initializer list

File: src/policy/fees.cpp [hunk 6]
```diff
@@ -956,22 +1002,36 @@ bool CBlockPolicyEstimator::Read(CAutoFile& filein)
...
-            // Copy buckets from file and refresh our bucketmap
-            buckets = fileBuckets;
-            bucketMap.clear();
-            for (unsigned int i = 0; i < buckets.size(); i++) {
-                bucketMap[buckets[i]] = i;
-            }
+LogPrintf("%s: loaded a bunch of fee estimation data, version %d, buckets %d (target: %d\n", __func__, nVersionThatWrote, numBuckets, buckets.size());
+
+LogPrintf("%s: buckets.size %d, buckets[0] %f, buckets[1] %f\n", __func__, buckets.size(), buckets[0], buckets[1]);
+LogPrintf("%s: fileBuckets.size %d, fileBuckets[0] %f, fileBuckets[1] %f\n", __func__, fileBuckets.size(), fileBuckets[0], fileBuckets[1]);
```
Comment (Empact) [inline src/policy/fees.cpp]: nit: whitespace

Comment (ajtowns) [reply src/policy/fees.cpp]: Ugh, those weren't meant to be committed in the first place...

File: src/test/mempool_tests.cpp (hunk 2 above)
Comment (luke-jr) [reply src/test/mempool_tests.cpp]: If tests don't already run in parallel, it would be nice if they did someday...

Comment (promag) [reply src/test/mempool_tests.cpp]: @luke-jr just curious, why? Running in parallel makes it harder to debug and reason about.

**Comment by promag** [comment]
Concept ACK.

Mind adding a release note?

**Comment by ajtowns** [comment]
Rebased and release note added

File: src/policy/fees.cpp (hunk 5 above)
Comment (ajtowns) [reply src/policy/fees.cpp]: this comment applies to #13990 now

File: doc/release-notes-13922.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+transactions across the network and for miners to consider the
+transaction in their blocks to 2 bits per kilobyte (0.00000200 BTC/kB,
+200 satoshis/kB), and the default additional fee rate required for
+replacing a transaction to 1 bit per kilobyte (0.00000100 BTC/kB, 100
+satoshis/kB). The previous default for all these values was 10 bits
+per kilobyte. These defaults can be overridden using the minrelaytxfee,
+blockmintxfee and incrementalrelayfee options.
+
+Note that until these lower defaults are widely adopted across the
+network, transactions created with lower fee rates may not propogate
```
Comment (practicalswift) [inline doc/release-notes-13922.md]: Typo found by `codespell`: propogate

File: test/functional/mempool_limit.py [hunk 8]
```diff
@@ -13,7 +13,7 @@ class MempoolLimitTest(BitcoinTestFramework):
     def set_test_params(self):
         self.setup_clean_chain = True
         self.num_nodes = 1
-        self.extra_args = [["-maxmempool=5", "-spendzeroconfchange=0"]]
+        self.extra_args = [["-maxmempool=5", "-spendzeroconfchange=0","-minrelaytxfee=0.00001"]]
```
Comment (practicalswift) [inline test/functional/mempool_limit.py]:
```
./test/functional/mempool_limit.py:16:70: E231 missing whitespace after ','
```

**Comment by ajtowns** [comment]
Rebased due to conflict in python tests; fixed formatting nit; fixed amount typo in changes to mempool test (7000/5 is 1400 not 1500).

**Comment by ajtowns** [comment]
Repeating some in-person comments:

 * reducing the min incremental relay fee allows for cheap spamming, might be better to change the code so that the min incremental relay fee can be higher than the min relay fee, and leaving it higher (perhaps 200s/kB for relay, but 1000s/kB or 500s/kB for incremental relay/fee-bumping). Note that wallet defaults already have a higher incremental fee, so this would improve consistency.

 * (long term, not for this PR) it might be nice if min incremental relay fees were dynamically determined based on how much traffic there is -- if there's no traffic, updating a tx doesn't need to cost anything significant; if there's lots of traffic, or a spam attack, updating a tx should become as expensive as issuing a new tx (ie, incremental relay fee should be as high as mempool minfee, which may be well above the min relay fee)

**Comment by ajtowns** [comment]
Closing for now pending further progress on #13990, intending to reopen or refile later

**Comment by brianddk** [comment]
@ajtowns It's been two years, and `minrelaytxfee` is still a lot higher than it probably needs to be.  Would you be willing to refile?  Someone needs to get this change through.

The mempool has been emptying pretty much every week, and often every day.  With LN interest catching on there is a larger need for "next-day" ultra-low fee on-chain TXNs.
